Presentation.Web/Controllers/API/ReferenceController.cs

```diff
@@ -41,35 +33,46 @@ protected override IControllerCrudAuthorization GetCrudAuthorization()
 
         public override HttpResponseMessage Post(ExternalReferenceDTO dto)
         {
-            if (dto.ItContract_Id.HasValue ||
-                dto.ItSystem_Id.HasValue ||
-                dto.ItSystemUsage_Id.HasValue ||
-                dto.ItProject_Id.HasValue)
+            if (dto == null)
             {
-                return base.Post(dto);
+                return BadRequest();
             }
-            return BadRequest("Target object must be specified");
+
+            return
+                GetOwnerTypeAndId(dto)
+                    .Match
+                    (
+                        onValue: typeAndId => _referenceService
+                            .AddReference
+                            (
+                                typeAndId.Value,
+                                typeAndId.Key,
+                                dto.Title,
+                                dto.ExternalReferenceId,
+                                dto.URL,
+                                dto.Display
+                            )
+                            .Match
+                            (
+                                onSuccess: NewObjectCreated,
+                                onFailure: FromOperationError
+                            ),
+                        onNone: () => BadRequest("Target owner Id must be defined")
+                    );
         }
 
-        protected override void PrepareNewObject(ExternalReference item)
+        private static Maybe<KeyValuePair<ReferenceRootType, int>> GetOwnerTypeAndId(ExternalReferenceDTO dto)
         {
-            if (item.ItProject_Id.HasValue)
-            {
-                item.ItProject = _projectRepository.GetById(item.ItProject_Id.Value);
-            }
-            if (item.Itcontract_Id.HasValue)
-            {
-                item.ItContract = _contractRepository.GetById(item.Itcontract_Id.Value);
-            }
-            if (item.ItSystem_Id.HasValue)
-            {
-                item.ItSystem = _systemRepository.GetSystem(item.ItSystem_Id.Value);
-            }
-            if (item.ItSystemUsage_Id.HasValue)
-            {
-                item.ItSystemUsage = _systemUsageRepository.GetSystemUsage(item.ItSystemUsage_Id.Value);
-            }
-            base.PrepareNewObject(item);
+            return GetOwnerTypeAndIdOrFallback(ReferenceRootType.Contract, dto.ItContract_Id,
+                fallback: () => GetOwnerTypeAndIdOrFallback(ReferenceRootType.System, dto.ItSystem_Id,
+                    fallback: () => GetOwnerTypeAndIdOrFallback(ReferenceRootType.SystemUsage, dto.ItSystemUsage_Id,
+                        fallback: () => GetOwnerTypeAndIdOrFallback(ReferenceRootType.Project, dto.ItProject_Id,
+                            fallback: () => Maybe<KeyValuePair<ReferenceRootType, int>>.None))));
```

Comment:
Hvad er fordelen her frem for en switch?

Reply:
Du kan kun switche på een værdi.. denne her switcher på 4 forskellige. I F# kunne man have switchet på en 4-val tuple med pattern matching